### ISSUE TITLE
SRT peer_idle_timeout not working on latest git

### ISSUE THREAD

**Issue by felipewd** [ossrs/srs#3390]
Hi,

I've been testing SRS as an SRT server. It works great!

However, on git-latest (`ef90da352`), I noticed if I start publishing an SRT stream, play it, and quit the `m=publish` side, it doesn't end the client even after the default 10s timeout. 

I then noticed the `peer_idle_timeout` param is not being recognized at all by the parser. That a simple fix with:

```
diff --git a/trunk/src/app/srs_app_config.cpp b/trunk/src/app/srs_app_config.cpp
index 89a15ee59..73ba00d1c 100644
--- a/trunk/src/app/srs_app_config.cpp
+++ b/trunk/src/app/srs_app_config.cpp
@@ -2313,7 +2313,7 @@ srs_error_t SrsConfig::check_normal_config()
             string n = conf->at(i)->name;
             if (n != "enabled" && n != "listen" && n != "maxbw"
                 && n != "mss" && n != "latency" && n != "recvlatency"
-                && n != "peerlatency" && n != "connect_timeout"
+                && n != "peerlatency" && n != "connect_timeout" && n != "peer_idle_timeout"
                 && n != "sendbuf" && n != "recvbuf" && n != "payloadsize"
                 && n != "default_app" && n != "sei_filter" && n != "mix_correct"
                 && n != "tlpktdrop" && n != "tsbpdmode" && n != "passphrase" && n != "pbkeylen") {
```

So now this works:

```
srt_server {
    enabled on;
    listen 1935;
    maxbw 1000000000;
    connect_timeout 4000;
    peer_idle_timeout 8000; # <== Accepted by the parser now
    peerlatency 2000;
    recvlatency 2000;
    latency 2000;
    tsbpdmode on;
    tlpktdrop on;
    sendbuf 2000000;
    recvbuf 2000000;
}
```

But the timeout itself doesn't seem to be working.

I do get this disposing of the sender-side:

```
[2023-01-27 19:29:47.546][INFO][23567][506bqbp7] SRT: disposing #0 resource(srt-ts-conn)(0x60d000006ab0), conns=2, disposing=1, zombies=0
[2023-01-27 19:29:47.546][INFO][23567][506bqbp7] close srt_fd=644381647
```

And little else after that.

The client-side (ffmpeg in my case) is "stuck" without getting a proper close/timeout on the stream.

Am I missing something on the config (built it based on `conf/srt.conf`) or is there a bug here?

Thanks, your code is awesome :-D

**Comment by winlinvip** [comment]
Seems a bug of configuration for SRT.

**Comment by felipewd** [comment]
hi @winlinvip thanks for the quick reply.

Not only a conf bug, because the simple patch above seems to fix that.

But the peer timeout itself seems to be completely ignored for SRT streams.

Since there's a 10s default value (I checked the code and it really is valid), if I start a `m=publish` stream and then a `m=request` (client) stream and then close the publish one, after a 10s timeout the request should close with an error. That doesn't happen. It just sits there, indefinitely.

So the bigger issue seems to be the timeout itself :-)

**Comment by xiaozhihong** [comment]
> hi @winlinvip thanks for the quick reply.
> 
> Not only a conf bug, because the simple patch above seems to fix that.
> 
> But the peer timeout itself seems to be completely ignored for SRT streams.
> 
> Since there's a 10s default value (I checked the code and it really is valid), if I start a `m=publish` stream and then a `m=request` (client) stream and then close the publish one, after a 10s timeout the request should close with an error. That doesn't happen. It just sits there, indefinitely.
> 
> So the bigger issue seems to be the timeout itself :-)

I got the problem. The libsrt update time when receiving data or ctrl message(ack/loss/...), and check expires in the timer.
But libsrt also send keepalive message every 1 second, so the peer_idle_timeout could no work.

**Comment by xiaozhihong** [comment]
I think you want to close all of the srt players when the publish closed, right?
It's so different with `peer_idle_timeout`, one is on the application side, other is on the transport side.

**Comment by felipewd** [comment]
hey @xiaozhihong, that's it!

If the publishing-side exited, I don't see a reason for the client-side just to hang there, not exiting ever.

Because I tested re-publishing after 10min on the same `streamid` but changing some major live avc -> hevc and the client-side just freaking out (correctly, of course) that the stream changed codecs "mid-playing" and exiting with an error. So not exiting becomes a major problem because we can't be sure if the publishing stream is just stalled, or if it has indeed been closed.

**Comment by xiaozhihong** [comment]
> hey @xiaozhihong, that's it!
> 
> If the publishing-side exited, I don't see a reason for the client-side just to hang there, not exiting ever.
> 
> Because I tested re-publishing after 10min on the same `streamid` but changing some major live avc -> hevc and the client-side just freaking out (correctly, of course) that the stream changed codecs "mid-playing" and exiting with an error. So not exiting becomes a major problem because we can't be sure if the publishing stream is just stalled, or if it has indeed been closed.

I make an [issue](https://github.com/Haivision/srt/issues/2653) to srt.
SRS has a [PR](https://github.com/ossrs/srs/pull/3105) about closing rtmp clients when the stream stop, this PR is similar to your requirement,  still in review.
Keep track of this issue, thanks.

**Comment by felipewd** [comment]
hey @xiaozhihong,

Given the comment here https://github.com/Haivision/srt/issues/2653#issuecomment-1429651602

I guess the proper behavior on SRS-side is to break the client-facing connection if there's no more publish? Ignoring the transport keep-alive, then.

Not sure if they'll do that on `libsrt` itself, but I'm thinking it makes sense to have that same behavior for SRT, RTMP and all other transport protocols for this problem, am I right?

The only question then becomes: How do I express an "idle publisher timeout" in order to close an inactive publisher, and notify all clients...if there's no clean-cut TCP-FIN equivalent for SRT.

**Comment by xiaozhihong** [comment]
> I guess the proper behavior on SRS-side is to break the client-facing connection if there's no more publish? Ignoring the transport keep-alive, then.

@felipewd 

You are right, but "close all players when publisher timeout or close", it's a specific requirement, maybe you can try to develop it yourselves. Implementing it in SRS needs discussion.

**Comment by Adam-Burke** [comment]
What is the status of this? I can't even seem to manually kick srt playback clients.
